File: dotnet/cs/ex_iis/ex_iis.cs
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////
////    LINDO API Version 8.0
////    Copyright (c; 2000-2014
////
////    LINDO Systems, Inc.            312.988.7422
////    1415 North Dayton St.          info@lindo.com
////    Chicago, IL 60622              http://www.lindo.com
////
////    @ex_iis.cs 
////
////    last updated: 07-25-2014
////
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Text;
using System.IO;

using System.Runtime.InteropServices;

public struct CallbackData
{
	public int calls;
}



public class ex_iis
{

    public static int n_cons=0;

	public static int n_vars=0;

    public static int MyCallbackLog(IntPtr pMod, string Msg, IntPtr myData)
    {
        int iter = 0;
        double pinf = 0, pobj = 0;

        // copy the user data in the unmanaged code into a local structure
        CallbackData cb = (CallbackData)Marshal.PtrToStructure(myData, typeof(CallbackData));

        // increment the number of calls to the callback function
        cb.calls++;

        Console.Write("{0}", Msg.ToString());

        // copy the user data in the local structure back to the unmanaged code
        Marshal.StructureToPtr(cb, myData, false);

        return 0;
    }

	public static int MyCallbackCount(IntPtr pMod, int nLoc,  IntPtr  myData)  
	{
		int iter=0;
		double pinf=0, pobj=0;   
      
		// copy the user data in the unmanaged code into a local structure
		CallbackData cb = (CallbackData) Marshal.PtrToStructure (myData, typeof(CallbackData));          
		
		// increment the number of calls to the callback function
		cb.calls++;     			
      
		lindo.LSgetCallbackInfo(pMod,0,lindo.LS_DINFO_PINFEAS, ref pinf);
		lindo.LSgetCallbackInfo(pMod,0,lindo.LS_DINFO_POBJ,ref pobj);
		lindo.LSgetCallbackInfo(pMod,0,lindo.LS_IINFO_SIM_ITER, ref iter);
		Console.WriteLine("callback @iter={0}, obj={1}, pinf={2}, ncalls={3}" ,iter,pobj,pinf,cb.calls );
            
		// copy the user data in the local structure back to the unmanaged code
		Marshal.StructureToPtr (cb,myData,false);

		return 0;
	}


    public static int MyCallback(IntPtr model, int loc, IntPtr nvCbData ) 
	{
        int it=0, nErr=lindo.LSERR_NO_ERROR;
        double ob=0.0;
        // get iterations
        nErr = lindo.LSgetCallbackInfo(model, loc, lindo.LS_IINFO_SIM_ITER, ref it);
        // get primal objective
        nErr = lindo.LSgetCallbackInfo(model, loc, lindo.LS_DINFO_POBJ, ref ob);

        Console.WriteLine("callback @iter={0}, obj={1}",it,ob);

        return 0;
    }

	public static void DisplayVersionInfo()
	{
        StringBuilder cMessage1 = new StringBuilder(lindo.LS_MAX_ERROR_MESSAGE_LENGTH);
        StringBuilder cMessage2 = new StringBuilder(lindo.LS_MAX_ERROR_MESSAGE_LENGTH);
        lindo.LSgetVersionInfo(cMessage1, cMessage2);
        Console.WriteLine(cMessage1);
        Console.WriteLine(cMessage2);
	}     


    public static void CheckErr(IntPtr env, int errorcode)
	{

        // Checks for an error condition.  If one exists, the
        //  error message is displayed then the application
        //  terminates.

        if (errorcode > 0) 
		{
            StringBuilder message = new StringBuilder(256);
            lindo.LSgetErrorMessage(env, errorcode, message);
            Console.WriteLine(message);
			return;
        }

    }
   
    public static void Main(string[] args)    
	{
        IntPtr env = (IntPtr) 0;
        IntPtr pModel = (IntPtr) 0;        
        int errorcode = lindo.LSERR_NO_ERROR;
		int nStatus = 0, customLogLevel=1;
		int i = 0;

		// number of constraints
		int m = 0;

		// number of variables
		int n = 0;

		// number of continous variables
		int ncont = 0;

		// objective value
        double obj = 0 ;        

        StringBuilder LicenseKey = new StringBuilder(lindo.LS_MAX_ERROR_MESSAGE_LENGTH);

        DisplayVersionInfo();
        if (args.Length != 1) 
		{
            Console.WriteLine("Usage: ex_iis filename");
            return;
        }


        // Read license key from file
        string LicenseFile = System.Environment.GetEnvironmentVariable("LINDOAPI_HOME") + "\\license\\lndapi140.lic";
        
        errorcode = lindo.LSloadLicenseString(LicenseFile, LicenseKey);
        if (errorcode > 0) { CheckErr(env, errorcode); return;}        

        // Create a LINDO environment.
        env = lindo.LScreateEnv(ref errorcode, LicenseKey.ToString());
        if (errorcode > 0) {
            Console.WriteLine("Unable to create environment.");
            return;
        }

        // Create a model in the environment.
        pModel = lindo.LScreateModel(env, ref errorcode);
        CheckErr(env, errorcode);

        // Optionally, declare user's callback data 
        // (total number of calls to the callback function)
        CallbackData cbData;
        cbData.calls = 0;
        IntPtr myData = Marshal.AllocHGlobal(Marshal.SizeOf(cbData));
        Marshal.StructureToPtr(cbData, myData, true);

        // Declare callback function
        lindo.typCallback cbFunc = new lindo.typCallback(ex_iis.MyCallbackCount);
        lindo.typModelLOG cbLog = new lindo.typModelLOG(ex_iis.MyCallbackLog);

        // Set callback or log function
        if (customLogLevel==3)
        {
            // set callback function and pass user data
            errorcode = lindo.LSsetCallback(pModel, cbFunc, myData);
        }
        else if (customLogLevel==1)
        {   // set log function and pass user data
            errorcode = lindo.LSsetModelLogfunc(pModel, cbLog, myData);
        }
        CheckErr(env, errorcode);

      
        // Read input model
        Console.WriteLine();        
        Console.WriteLine("Reading {0}.",args[0]);
        errorcode = lindo.LSreadMPSFile( pModel, args[0], lindo.LS_UNFORMATTED_MPS);
        if (errorcode>0) {
        	errorcode = lindo.LSreadLINDOFile( pModel, args[0]);       
        }
        CheckErr(env, errorcode);

		// get model stats
		errorcode = lindo.LSgetInfo(pModel, lindo.LS_IINFO_NUM_VARS, ref n);
		errorcode = lindo.LSgetInfo(pModel, lindo.LS_IINFO_NUM_CONS, ref m);
		errorcode = lindo.LSgetInfo(pModel, lindo.LS_IINFO_NUM_CONT, ref ncont);

		ex_iis.n_vars = n;
		ex_iis.n_cons = m;

		double [] x = new double[n];
		double [] y = new double[m];

		// Perform the optimization.
		Console.WriteLine("Solving...");
        if (n==ncont) 
            errorcode = lindo.LSoptimize(pModel, lindo.LS_METHOD_FREE, ref nStatus);
        else
            errorcode = lindo.LSsolveMIP(pModel, ref nStatus);

		CheckErr(env, errorcode);
        if (nStatus == lindo.LS_STATUS_BASIC_OPTIMAL || nStatus == lindo.LS_STATUS_OPTIMAL)
        {
            Console.WriteLine("\tThe model is solved to optimality.\n");
            // Retrieve the solution and print
            if (n == ncont)
            {
                errorcode = lindo.LSgetInfo(pModel, lindo.LS_DINFO_POBJ, ref obj);
                CheckErr(env, errorcode);
                errorcode = lindo.LSgetPrimalSolution(pModel, x);
                CheckErr(env, errorcode);
                errorcode = lindo.LSgetDualSolution(pModel, y);
                CheckErr(env, errorcode);
            }
            else
            {                    
                errorcode = lindo.LSgetInfo(pModel, lindo.LS_DINFO_MIP_OBJ, ref obj);
                CheckErr(env, errorcode);
                errorcode = lindo.LSgetMIPPrimalSolution(pModel, x);
                CheckErr(env, errorcode);
                errorcode = lindo.LSgetMIPDualSolution(pModel, y);
                CheckErr(env, errorcode);
            }
            Console.WriteLine();
            Console.WriteLine("Objective value = {0} ", obj);         

        }
       /***************************************************************
        * Step 6: Debug the model if unbounded or infeasible
        ***************************************************************/
        else if (nStatus == lindo.LS_STATUS_UNBOUNDED)
        {
            Console.WriteLine("\nThe model is unbounded.. Analyzing...\n");
            int nLevel = lindo.LS_NECESSARY_COLS + lindo.LS_SUFFICIENT_COLS;

            /*** Step 6.1: Find IIS ***/
            errorcode = lindo.LSfindIUS(pModel, nLevel);
            CheckErr(env, errorcode);
        }
        else if (nStatus == lindo.LS_STATUS_INFEASIBLE)
        {
            Console.WriteLine("\nThe model is infeasible.. Analyzing...\n");
            int nLevel = lindo.LS_NECESSARY_ROWS + lindo.LS_SUFFICIENT_ROWS;
            int nSuf_r = 0, nIIS_r = 0, nSuf_c = 0, nIIS_c = 0;
            int[] aiRows = new int[m];
            int[] aiCols = new int[n];
            int[] anBnds = new int[n];

            /*** Step 6.1: Find IIS ***/
            errorcode = lindo.LSfindIIS(pModel, nLevel);
            CheckErr(env, errorcode);

            errorcode = lindo.LSgetIIS(pModel, ref nSuf_r, ref nIIS_r, aiRows, ref nSuf_c, ref nIIS_c, aiCols, anBnds);
            CheckErr(env, errorcode);
            StringBuilder varname = new StringBuilder(256);
            String bndtype;

            if (customLogLevel == 2)
            {
                Console.WriteLine("\n\t ***  LSfindIIS Summary ***\n");
                Console.WriteLine("\t Number of Sufficient Rows = {0}", nSuf_r);
                Console.WriteLine("\t Number of Sufficient Cols = {0}", nSuf_c);
                Console.WriteLine("\t Number of Necessary  Rows = {0}", nIIS_r - nSuf_r);
                Console.WriteLine("\t Number of Necessary  Cols = {0}", nIIS_c - nSuf_c);
                Console.WriteLine("\n");

                /*** Step 6.2: Display row index sets ***/
                Console.WriteLine("\n IIS Rows");
                for (int j = 0; j < nIIS_r; j++)
                {
                    errorcode = lindo.LSgetConstraintNamei(pModel, aiRows[j], varname);
                    CheckErr(env, errorcode);

                    if (j < nSuf_r)
                        Console.WriteLine("{0}] ({1}) is in the sufficient set.\n", j, varname);
                    else
                        Console.WriteLine("{0}] ({1}) is in the necessary set.\n", j, varname);
                }//for-j

                /*** Step 6.3: Display column index sets ***/
                Console.WriteLine("\n IIS Column Bounds\n");
                for (int j = 0; j < nIIS_c; j++)
                {
                    if (anBnds[j] < 0)
                        bndtype = "Lower";
                    else
                        bndtype = "Upper";

                    errorcode = lindo.LSgetVariableNamej(pModel, aiCols[j], varname);
                    CheckErr(env, errorcode);
                    if (j < nSuf_r)
                        Console.WriteLine("{0}] %s bound of ({1}) is in the sufficient set.\n", j, bndtype, varname);
                    else
                        Console.WriteLine("{0}] %s bound of ({1}) is in the necessary set.\n", j, bndtype, varname);
                }//for-j
            }
        }//if
        
        Console.WriteLine();
	    cbData = (CallbackData) Marshal.PtrToStructure(myData,typeof(CallbackData));
        Console.WriteLine("Total callbacks : {0}" ,cbData.calls );
               
		// free user data in global heap
        Marshal.FreeHGlobal(myData);           

        // Delete the LINDO environment.
        lindo.LSdeleteEnv(ref env);

		Console.WriteLine("Press Enter..."); 
		Console.ReadLine();

     }
}

File: c/ex_iis/ex_iis.c
/*
###################################################################
#                       LINDO-API
#                    Sample Programs
#
#                  Copyright (c) 2011
#
#         LINDO Systems, Inc.           312.988.7422
#         1415 North Dayton St.         info@lindo.com
#         Chicago, IL 60622             http://www.lindo.com
###################################################################

  File   : ex_iis.c

  Purpose: Analyze an infeasible (unbounded) LP to isolate the
  constraints (variables) causing the infeasibility (unboundedness)
  of the model.
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

/* LINDO API header file */
#include "lindo.h"


/* Define a macro to declare variables for
    error checking */
#define APIERRORSETUP  \
   int nErrorCode; \
   char cErrorMessage[LS_MAX_ERROR_MESSAGE_LENGTH] \

/* Define a macro to do our error checking */
#define APIERRORCHECK  \
   if (nErrorCode) \
   { \
      if ( pEnv) \
      { \
         LSgetErrorMessage( pEnv, nErrorCode, \
          cErrorMessage); \
         printf("Errorcode=%d:  %s\n", nErrorCode, \
          cErrorMessage); \
      } else {\
         printf( "Fatal Error\n"); \
      } \
      goto Terminate; \
   } \

#define APIVERSION \
{\
    char szVersion[255], szBuild[255];\
    LSgetVersionInfo(szVersion,szBuild);\
    printf("\nLINDO API Version %s built on %s\n",szVersion,szBuild);\
}\

int CALLBACKTYPE MyCallback( pLSmodel pMod, int nLocation,
 void* pMyData)
{
   int *status = (int *) pMyData;

/* Display current iteration count and objective value */
   {
      int nIter,nNec=0,nSuf=0,
         nErr,nErr1,nErr2;
      double dObj, dInf;
      nErr=LSgetCallbackInfo(pMod,nLocation,LS_IINFO_SIM_ITER,&nIter);
      nErr=LSgetCallbackInfo(pMod,nLocation,LS_DINFO_POBJ,&dObj);
      nErr=LSgetCallbackInfo(pMod,nLocation,LS_DINFO_PINFEAS,&dInf);
      if (status && *status == LS_STATUS_INFEASIBLE)
      {
        nErr1=LSgetCallbackInfo(pMod,nLocation,LS_IINFO_NUM_IIS_ROWS,&nNec);
        nErr2=LSgetCallbackInfo(pMod,nLocation,LS_IINFO_NUM_SUF_ROWS,&nSuf);
      }
      else if (status && *status == LS_STATUS_UNBOUNDED)
      {
        nErr1=LSgetCallbackInfo(pMod,nLocation,LS_IINFO_NUM_IUS_COLS,&nNec);
        nErr2=LSgetCallbackInfo(pMod,nLocation,LS_IINFO_NUM_SUF_COLS,&nSuf);
      }

      printf( "@MyCallback %8d, %8d, %16g, %16g, %8d (%d)\n",
        nLocation,nIter,dObj,dInf,nNec,nSuf);

   }

   return( 0);
}

static void LS_CALLTYPE print_line_log(pLSmodel pModel, char *line, void *userdata)
{
  if (line)
  {
    printf("\n%s",line);
  } /*if*/
} /*print_line*/

 /* main entry point */
int main(int argc, char **argv)
{
   APIERRORSETUP;
   /* model data objects */
   int n; /* number of variables */
   int m; /* number of constraints */
   int solstatus;/*solution status (see lindo.h for possible values)*/
   int prep_level;
   char *mpsfile = NULL;
   char MY_LICENSE_KEY[1024];

   /* IIS related data objects */
   int nLevel,   /* level of analysis */
       nSuf_r,   /* number of sufficient rows     */
       nSuf_c,   /* number of sufficient columns  */
       nIIS_r,   /* number of rows in the IIS     */
       nIIS_c;   /* number of columns in the IIS  */
   int *aiRows = NULL, /* index set of rows in the IIS     */
       *aiCols = NULL, /* index set of columns in the IIS  */
       *anBnds = NULL; /* bound type of columns in the IIS */
   int j;
   char bndtype[255], oufname[255], varname[255];

   /* declare an instance of the LINDO environment object */
   pLSenv pEnv=NULL;
   /* declare an instance of the LINDO model object */
   pLSmodel pModel;
   /***************************************************************
    * Init: Command prompt calling sequence
    ***************************************************************/
   {
     char szVer[255], szBld[255];
     LSgetVersionInfo(szVer,szBld);
     printf("\nAN APPLICATION FOR ANALYZING & DEBUGGING LPs\n");
     printf("\nusing LINDO API Version %s (Built %s)\n\n",szVer,szBld);

   }
   if (argc == 1)
   {
     printf("\nUsage: ex_iis filename\n\n");
     goto Terminate;
   }
   else if (argc == 2)
   {
     mpsfile = argv[1];
   }
  /*****************************************************************
   * Step 1: Create a LINDO environment.
   *****************************************************************/
   nErrorCode = LSloadLicenseString("../../../license/lndapi130.lic",MY_LICENSE_KEY);
   if ( nErrorCode != LSERR_NO_ERROR)
   {
      printf( "Failed to load license key (error %d)\n",nErrorCode);
      exit( 1);
   }

   APIVERSION;
   pEnv = LScreateEnv ( &nErrorCode, MY_LICENSE_KEY);
   if ( nErrorCode == LSERR_NO_VALID_LICENSE)
   {
      printf( "Invalid License Key!\n");
      exit( 1);
   }
   APIERRORCHECK;

   /***************************************************************
    * Step 2: Create a model in the environment.
    ***************************************************************/
   pModel = LScreateModel ( pEnv, &nErrorCode);
   APIERRORCHECK;

   /***************************************************************
    * Step 3: Read the model from an MPS file and
    ***************************************************************/
   nErrorCode = LSreadMPSFile(pModel,mpsfile,LS_UNFORMATTED_MPS);
   if (nErrorCode != LSERR_NO_ERROR)
   {
     printf("\nBad MPS format... Trying LINDO format.\n");
     nErrorCode =LSreadLINDOFile(pModel,mpsfile);
     APIERRORCHECK;
     printf("\nLINDO format OK!\n\n");
   }
   else
   {
     printf("\nMPS format OK!\n\n");
   }
   nErrorCode = LSgetInfo(pModel, LS_IINFO_NUM_VARS, &n);
   APIERRORCHECK;

   nErrorCode = LSgetInfo(pModel, LS_IINFO_NUM_CONS, &m);
   APIERRORCHECK;

   /***************************************************************
    * Step 4: Set Model parameters
    ***************************************************************/
   /* Turn off the LP preprocessor. This is required if the model
   is infeasible and the user wishes to debug it. */
   nErrorCode = LSgetModelIntParameter(pModel,LS_IPARAM_LP_PRELEVEL,
     &prep_level);
   APIERRORCHECK;
   if (prep_level > 0)
     printf("The LP presolver has been turned off. Solving ...\n\n");

   nErrorCode = LSsetModelIntParameter(pModel,LS_IPARAM_LP_PRELEVEL,0);

   /* set LP solver type for optimizations (cold start) */
   nErrorCode = LSsetModelIntParameter(pModel,LS_IPARAM_IIS_TOPOPT,
     LS_METHOD_FREE);

   /* set LP solver type for reoptimizations (warm start)*/
   nErrorCode = LSsetModelIntParameter(pModel,LS_IPARAM_IIS_REOPT,
     LS_METHOD_FREE);

#if 0
   nErrorCode = LSsetCallback( pModel,(cbFunc_t) MyCallback, NULL);
   APIERRORCHECK;
   printf( "            %8s, %8s, %16s, %16s, %8s (%s)\n",
        "LOCATION","ITERS","OBJECTIVE","INFEASIBILITY","NNEC","NSUF");
#endif

   /* Install a log function to display solver's progress
   as reported by the internal solver */
   nErrorCode = LSsetModelLogfunc(pModel, (printModelLOG_t) print_line_log, NULL);

   nErrorCode = LSsetModelDouParameter(pModel,LS_DPARAM_CALLBACKFREQ,0.5);
   APIERRORCHECK;

   /***************************************************************
    * Step 5: Optimize the model
    ***************************************************************/
   nErrorCode = LSoptimize( pModel,LS_METHOD_FREE, &solstatus);
   APIERRORCHECK;

#if 0
   /* set callback and solution status */
   nErrorCode = LSsetCallback( pModel,(cbFunc_t) MyCallback, &solstatus);
#endif

    if (solstatus == LS_STATUS_BASIC_OPTIMAL)
    {
      printf("\tThe model is solved to optimality.\n");
    }
   /***************************************************************
    * Step 6: Debug the model if unbounded or infeasible
    ***************************************************************/
    else if (solstatus == LS_STATUS_UNBOUNDED)
    {
      APIERRORCHECK;
      printf("\nThe model is unbounded.. Analyzing...\n\n");
      nLevel = LS_NECESSARY_COLS + LS_SUFFICIENT_COLS;

      /*** Step 6.1: Find IIS ***/
      nErrorCode = LSfindIUS(pModel,nLevel);
      APIERRORCHECK;

      strcpy(oufname,"findius.ltx");
      nErrorCode = LSwriteIUS(pModel,oufname);
      printf("\n\n IUS is written to %s !!\n",oufname);
    }
    else if (solstatus == LS_STATUS_INFEASIBLE)
    {
      printf("\nThe model is infeasible.. Analyzing...\n\n");
      aiRows = (int *) malloc(m*sizeof(int));
      aiCols = (int *) malloc(n*sizeof(int));
      anBnds = (int *) malloc(n*sizeof(int));

      /*** Step 6.1: Find IIS ***/
      nLevel = LS_NECESSARY_ROWS + LS_SUFFICIENT_ROWS;

      nErrorCode = LSfindIIS(pModel,nLevel);
      APIERRORCHECK;

      nErrorCode = LSgetIIS(pModel,&nSuf_r,&nIIS_r,aiRows,
                                   &nSuf_c,&nIIS_c,aiCols,anBnds);
      APIERRORCHECK;
      printf("\n\t ***  LSfindIIS Summary ***\n\n");
      printf("\t Number of Sufficient Rows = %u\n",nSuf_r);
      printf("\t Number of Sufficient Cols = %u\n",nSuf_c);
      printf("\t Number of Necessary  Rows = %u\n",nIIS_r - nSuf_r);
      printf("\t Number of Necessary  Cols = %u\n",nIIS_c - nSuf_c);
      printf("\n");

      /*** Step 6.2: Display row index sets ***/
      printf("\n IIS Rows\n");
      for (j=0; j<nIIS_r; j++)
      {
        nErrorCode = LSgetConstraintNamei(pModel,aiRows[j],varname);
        APIERRORCHECK;

        if (j<nSuf_r)
          printf("%2d] (%-8s) is"
          " in the sufficient set.\n",j,varname);
        else
          printf("%2d] (%-8s) is"
          " in the necessary set.\n",j,varname);
      }

      /*** Step 6.3: Display column index sets ***/
      printf("\n IIS Column Bounds\n");
      for (j=0; j<nIIS_c; j++)
      {
        if (anBnds[j] < 0)
          strcpy(bndtype,"Lower");
        else
          strcpy(bndtype,"Upper");

        nErrorCode = LSgetVariableNamej(pModel,aiCols[j],varname);
        APIERRORCHECK;
        if (j<nSuf_c)
          printf("%2d] %s bound of (%-8s) is"
          " in the sufficient set.\n",j,bndtype,varname);
        else
          printf("%2d] %s bound of (%-8s) is"
          " in the necessary set.\n",j,bndtype,varname);
      }

      strcpy(oufname,"findiis.ltx");
      LSwriteIIS(pModel,oufname);
      printf("\n\n IIS is written to %s !!\n",oufname);

      free(aiRows);
      free(aiCols);
      free(anBnds);
    }

Terminate:
 /*****************************************************************
  * Step 7: Terminate
  *****************************************************************/
   nErrorCode = LSdeleteModel( &pModel);

   nErrorCode = LSdeleteEnv( &pEnv);

  /* Wait until user presses the Enter key */
   printf("Press <Enter> ...");
   getchar();


}
